### DISCUSSION TITLE
Version 1.0.0 release notes

### DISCUSSION THREAD

**Discussion by lemire** [simdjson/simdjson#1710] [General]
I plan to deploy release 1.0 this week unless some issue comes up. I'd like feedback on the release notes.

----------

Release 1.0.0 of the simdjson library builds on earlier pre-1.0 release that made the On Demand frontend our default. The On Demand front-end is a new way to build parser. With On Demand, if you open a file containing 1000 numbers and you need just one of these numbers, only one number is parsed. If you need to put the numbers into your own data structure, they are materialized there directly, without being first written to a temporary tree. Thus we expect that the simdjson On Demand might often provide superior performance.

Release 1.0.0 adds several key features:

- In big data analytics, it is common to serialize large sets of records as multiple JSON documents separated by while spaces.  You can now get the benefits of On Demand while parsing almost infinitely long streams of JSON records (see iterate_many). At each step, you have access to the current document, but a secondary thread indexes the following block. You can thus access enormous files while using a small amount of memory and achieve record-breaking speeds.
- In some cases, JSON documents contain numbers embedded within strings (e.g., "3.1416"). You can access these numbers directly using methods such as get_double_in_string() .
- Given an On Demand instance (value, array, object, etc.), you can now convert it to a JSON string using the to_json_string method which returns a string view in the original document for unbeatable speeds.
- The On Demand front-end now supports the JSON Pointer specification. You can request a specific value using a JSON Pointer within a large document.
- Arrays in On Demand now have a count_elements() method. Objects have a count_fields() method. Arrays and objects have a reset method for when you need to iterate through them more than once. Document instances now have a rewind method in case you need to process the same document multiple times.

Other improvements include:

- We have extended and improved our documentation and we have added much testing.
- We have accelerated the JSON minification function (`simdjson::minify`) under ARM processors (credit @dougallj) 

We encourage users of previous versions of the simdjson library to update. We encourage users to deploy it for production uses.

The On Demand front-end was invented by @jkeiser. Much of the work of release 1.0 should be credited to @NicolasJiaxin.

**Comment by JeffGarland** [comment]
Thanks for deploying release 1.0 and the great library in general.  I'll make the comment here since there seems to be a small inconsistency in the 1.0 docs.  Specifically in the on-demand limitations section on this page

https://simdjson.org/api/1.0.0/md_doc_ondemand_design.html

The last bullet is:

- The On Demand API does not support JSON Pointer. This capability is currently limited to our core API.

Which I'm guessing is now out of date based on the bullet above.  I've also been using at_pointer for something I'm doing with an  ondemand parser and it's working great.

**Reply by lemire** [reply]
Correct!!!

**Comment by ghost** [comment]
Is all the sample code out of date?

I'm just trying to iterate directory over the key-value pairs, look at the key name and decode from that, about the simplest streaming case that can be done, but all the references to `kv.key` and `kv.value` seem to have changed to functions and hidden the actual string somewhere. No escaped keys, I just need to do a quick string match so getting the string_view would be nice. The casting the value after figure out what key. (posting here because I cannot start a new topic). I just copied the single header and cpp from master before this.

**Reply by lemire** [reply]
You refer to outdated code examples, can you be specific? Our code examples are tested and should work. Errors are always possible, of course… but we’ve had no such report so far.

**Reply by lemire** [reply]
We have many related examples in our documentation… see…

https://github.com/simdjson/simdjson/blob/master/doc/basics.md#error-handling-examples-without-exceptions

**Reply by lemire** [reply]
Note that we have two distinct frontends… a conventional DOM and On Demand. Maybe your code was built using DOM. It is still supported but no longer our default.

**Reply by ghost** [reply]
Do you have an example of a stream parser? I don't have the entire document yet (and probably never will have more than a couple TCP frame in memory at once but each object is relatively flat and small except a couple top level arrays).

The code I'm referring to is anything that does `(auto kv : parser.iterate (buf, len, len+PADDING).get_object()) { ... kv.key ... }` none of it compiles for me and I need to do `kv.key().raw().value()` but that seems to be a c-string and not the string_view I was expecting. This is done with `ondemand::*`. The example code all treat `key` as a variable, but the single-file version I have treat it as a method. Unless I'm doing something horribly wrong, which is very likely.

Here's a small piece of code:

```
        ondemand::document doc = parser.iterate (buf->read_head (), len, len + PAD);
        ondemand::object   obj = doc.get_object ();
        for (auto e : obj) {
            char const *k = e.key ().raw ().value ();
            switch (*k) {
            case 'x':
                assert(k[1] == '0');
                t.time = e.value ().get_int64 ();
```

**Reply by lemire** [reply]
An object is a collection of fields in On Demand. 

https://github.com/simdjson/simdjson/blob/95b4870e20be5f97d9dcf63b23b1c6f520c366c1/include/simdjson/generic/ondemand/field.h#L14-L44

A field is made of a raw_json_string and a value. The raw_json_string is a simple pointer to the location of the key...


https://github.com/simdjson/simdjson/blob/2abcc35031152ac4a8d867775b035203a8ae7541/include/simdjson/generic/ondemand/raw_json_string.h


If you want to check that the key is equal to the string `"x"`, you should be able to do...

```
if (e.key() == "x") {
...
}
```

In you example, you want to branch on just the first character... you can certain do that with `raw()`, it will work.

For streaming use, please refer to our documentation: 

https://github.com/simdjson/simdjson/blob/master/doc/iterate_many.md

As for the rest of the issues, if you believe that there is a bug, please consider writing a unit test and filing it as an issue.
